### PR TITLE
Github Actions: improvements to CI workflows

### PR DESCRIPTION
**Pull request by alalazo**
Closes #17754
Closes #17738

Modifications:
- [x] Group tests with similar duration together
- [x] Slightly simplify YAML files for workflows
- [x] Fetch all branches and tags from spack repo using `fetch-depth: 0`

**Group tests with similar duration together:** style and documentation tests take just a few minutes to run. Since in Github actions one can't restart a single job but needs to restart an entire workflow, here we group tests with similar duration together. The benefit is that if a test lasting 1 min. is hanging we don't need to restart a few other ones that would take much longer to complete.

**Fetch all branches and tags:** MacOS unit tests started failing as the MacOS virtual environment was upgraded to use git 2.28.0 The mechanism leading to the failure is well explained in #17754 description. This PR solves the issue by fetching the whole history from Spack repository and by using `git branch develop origin/develop` instead of `git fetch -u origin develop:develop`.

### CODE REVIEW

File: .github/workflows/setup_git.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env sh
+git config --global user.email "spack@example.com"
+git config --global user.name "Test User"
+# With fetch-depth: 0 we have a remote develop
+# but not a local branch. Don't do this on develop
+if [ "$(git branch --show-current)" != "develop" ]
+then
+  git branch develop origin/develop
+fi
```
Comment (alalazo) [inline .github/workflows/setup_git.sh:9]: To reviewers: please double check that this will prevent us from doing `git branch develop origin/develop` while currently on `develop`.

File: .github/workflows/linux_build_tests.yaml [hunk 2]
```diff
@@ -23,23 +23,19 @@ on:
 
 jobs:
   build:
-
     runs-on: ubuntu-latest
     strategy:
-      max-parallel: 4
```
Comment (scheibelp) [inline .github/workflows/linux_build_tests.yaml:29]: Why was this removed?

File: .github/workflows/style_and_docs.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+on:
+  push:
+    branches:
+      - develop
+      - releases/**
+  pull_request:
+    branches:
+      - develop
+      - releases/**
```
Comment (scheibelp) [inline .github/workflows/style_and_docs.yaml:11]: The doc/flake8 checks don't run in all the cases that they used to - from `linux_unit_tests.yaml`:

```

  push:
    branches:
      - master
      - develop
      - releases/**
  pull_request:
    branches:
      - master
      - develop
```

I think it may make sense to keep these conditions the same, so perhaps the version check could remain in a separate file?

File: .github/workflows/setup_git.sh (hunk 1 above)
Comment (scheibelp) [reply .github/workflows/setup_git.sh:9]: `--show-current` is supported for 2.22 and up. So this looks fine (including the `sh` syntax).

When do you expect this to be false? I assume that if we were to `git branch develop origin/develop` when there was already a develop branch, the command would fail. Would it also be possible to do this with `-f` like `git branch -f develop origin/develop`?

File: .github/workflows/linux_build_tests.yaml [hunk 4]
```diff
@@ -23,23 +23,19 @@ on:
...
       matrix:
         package: [lz4, mpich, tut, py-setuptools, openjpeg, r-rcpp]
     steps:
     - uses: actions/checkout@v2
-    - name: Cache ccache's store
-      uses: actions/cache@v1
+    - uses: actions/cache@v2
```
Comment (scheibelp) [inline .github/workflows/linux_build_tests.yaml:32]: Are these version updates necessary to achieve the effects mentioned in the PR description?

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a few questions and a request

File: .github/workflows/linux_build_tests.yaml (hunk 2 above)
Comment (alalazo) [reply .github/workflows/linux_build_tests.yaml:29]: I think this was a leftover, so I removed it. We removed custom limitations in all the other files and are currently relying on Github's default.

File: .github/workflows/style_and_docs.yaml (hunk 3 above)
Comment (alalazo) [reply .github/workflows/style_and_docs.yaml:11]: We removed the `master` branch recently so there is no need for that run. The current workflow adds a new trigger (run on pull requests for branches that match `releases/**`)

File: .github/workflows/setup_git.sh (hunk 1 above)
Comment (alalazo) [reply .github/workflows/setup_git.sh:9]: > When do you expect this to be false?

This will be false as soon as we merge a PR and retrigger CI on a push to `develop`. 

> I assume that if we were to git branch develop origin/develop when there was already a develop branch, the command would fail

Indeed the command fails with non-zero exit code, and that would cause the entire workflow to fail.

File: .github/workflows/linux_build_tests.yaml (hunk 4 above)
Comment (alalazo) [reply .github/workflows/linux_build_tests.yaml:32]: No, these are just "maintenance" updates to the actions we use - since I was at it I just bumped versions of all the ones that were lagging behind the latest stable release.

**Review by scheibelp — APPROVED** [review]
